### PR TITLE
Add support for rand_like op in fusion compiler

### PR DESCRIPTION
**Pull request by runtian-zhou**
Enabled support for generating random numbers in fusion compiler. Currently a philox RNG implemented by Tensorflow is used, as the NVRTC couldn't resolve the curand.h header correctly. The two implementation should have exact same behavior according to our tests.

### CODE REVIEW

**Comment by ezyang** [comment]
needs rebasing, sorry

File: torch/csrc/jit/fusion_compiler.cpp [hunk 1]
```diff
@@ -80,12 +84,297 @@ struct TensorInfo {
   IndexType strides[N];
 };
 )");
+constexpr auto rand_support_literal = R"(
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: It would be very good to mention in a comment above this WHY curand can't be used, so it's obvious to anyone reading this code. (IIUC, it's because nvrtc doesn't support curand, is that right? So if one day, nvrtc does support curand, we know we can get rid of this header.)

File: torch/csrc/jit/fusion_compiler.cpp [hunk 2]
```diff
@@ -80,12 +84,297 @@ struct TensorInfo {
...
+constexpr auto rand_init = R"(
+  int idx = blockIdx.x*blockDim.x + threadIdx.x;
+  PhiloxWrapper rnd(seed, idx, offset);
+)";
 auto cuda_compilation_unit_template = CodeTemplate(R"(
 ${type_declarations}
 
 extern "C" __global__
-void ${kernelName}(IndexType totalElements, ${formals}) {
+void ${kernelName}(IndexType totalElements, ${formals} ${RandParam}) {
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Nit: our template engine supports automatically inserting the comma if the parameter is non-null, using the syntax `${,RandParam}`. So you don't need to put the comma in `rand_param` above.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 3]
```diff
@@ -555,6 +862,14 @@ void CompiledFusionFunction::launch_with_tensors(at::ArrayRef<at::Tensor> inputs
       }
     }
   }
+  // If the kernel call contains a random op, we need to pass in random seeds as
+  // well.
+  if(has_random) {
+    auto gen_ = THCRandom_getGenerator(at::globalContext().getTHCState());
+    uint64_t offset = gen_->state.philox_seed_offset.fetch_add(20);
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: 20? That seems... very questionable :)

File: torch/csrc/jit/fusion_compiler.cpp [hunk 4]
```diff
@@ -590,13 +905,13 @@ struct CUDAFusionFunction : public CompiledFusionFunction {
...
     std::stringstream cu;
-    concat_desc = codegen::emitCompilationUnit(cu, name, agraph, true);
+    concat_desc = codegen::emitCompilationUnit(cu, has_random, name, agraph, true);
     compilation_unit = cu.str();
     nvrtcProgram program;
     TORCH_NVRTC_CHECK(nvrtcCreateProgram(&program, compilation_unit.c_str(), NULL, 0, nullptr, nullptr));
 
     std::string compute = "--gpu-architecture=compute_" + std::to_string(prop.major) + std::to_string(prop.minor);
-    std::vector<const char *> args = {"--std=c++11", compute.c_str()};
+    std::vector<const char *> args = {"--std=c++11", compute.c_str(), "-default-device"};
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: Why can't we just annotate our entities with `__device__` as necessary? Is this because of the TF Philox code?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 5]
```diff
@@ -310,9 +600,11 @@ std::string encodeRHS(Node * n) {
 }
 
 std::vector<ConcatDesc> emitCompilationUnit(std::ostream & out,
+                                            bool& has_random,
```
Comment (ezyang) [inline torch/csrc/jit/fusion_compiler.cpp]: This is counter to FB/Google C++ style. Pass in a `bool*` instead.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 3 above)
Comment (ezyang) [reply torch/csrc/jit/fusion_compiler.cpp]: At compilation time, don't we know how many samples we need to take? So you can just save it in the compilation unit and then grab exactly the correct offset.

**Review by ezyang — APPROVED** [review]
This is a really tight PR, I like it a lot! Good to go after the nits are fixed.

File: torch/csrc/jit/fusion_compiler.cpp (hunk 4 above)
Comment (zrt95) [reply torch/csrc/jit/fusion_compiler.cpp]: Right

File: torch/csrc/jit/fusion_compiler.cpp (hunk 3 above)
Comment (zrt95) [reply torch/csrc/jit/fusion_compiler.cpp]: Actually I wasn't quite sure what numbers to put here. I was looking at some previous usage of philox seed in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Distributions.cu and I just saw those random numbers are being used here.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 6]
```diff
@@ -566,6 +882,18 @@ void CompiledFusionFunction::launch_with_tensors(at::ArrayRef<at::Tensor> inputs
       }
     }
   }
+
+  #ifdef USE_CUDA
+    // If the kernel call contains a random op, we need to pass in random seeds as
+    // well.
+    if(has_random && this->backend() == at::kCUDA) {
+      auto gen_ = THCRandom_getGenerator(at::globalContext().getTHCState());
+      uint64_t offset = gen_->state.philox_seed_offset.fetch_add(20);
```
Comment (zdevito) [inline torch/csrc/jit/fusion_compiler.cpp]: Can you explain where this `20` comes from and why it is correct?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 7]
```diff
@@ -321,9 +619,12 @@ std::string encodeRHS(Node * n) {
 }
 
 std::vector<ConcatDesc> emitCompilationUnit(std::ostream & out,
+                                            bool* has_random,
```
Comment (zdevito) [inline torch/csrc/jit/fusion_compiler.cpp]: nit: we typically put return values in returns:

```
std::pair<std::vector<ConcatDesc>, bool> emitCompilationUnit
```

**Review by zdevito — COMMENTED** [review]
Structure looks good. I have a few questions about how random number seeds are handled. For instance, if we have two calls to rand inside a kernel, it doesn't look like we change the amount we increment the seed accordingly. Can you explain how this works?

File: torch/csrc/jit/fusion_compiler.cpp (hunk 6 above)
Comment (zrt95) [reply torch/csrc/jit/fusion_compiler.cpp]: Actually I wasn't quite sure what to put here as well. The reference I had is pytorch/aten/src/ATen/native/cuda/Distributions.cu. When I talked to other developers about this, they told me "10 or 20 is a rough estimate on how many times the generator will be called. It's not ideal, but authors of those functions decided that it is good enough.".

Comment (zdevito) [reply torch/csrc/jit/fusion_compiler.cpp]: Unlike the other uses, the number of times an inner loops will run in a fusion kernel is dependent on the size of the tensors and the number of calls to rand.

In that case shouldn't it be  (number of times a single thread can call the random number generator)/4?

Max times a thread runs its inner loop is:

```
max_times_an_inner_loop_runs = ceil(numel/(num_threads_per_block*number_of_blocks_launched))
max_rand_calls_per_thread = num_calls_to_rand_in_body*max_times_an_inner_loop_runs
```

Comment (ngimel) [reply torch/csrc/jit/fusion_compiler.cpp]: It shouldn't be divided by 4, I think. Philox counter the way it's implemented here is counting every number. See in my pr for offset increment https://github.com/pytorch/pytorch/pull/9666/files#diff-13a03e6a295e68ffd7ca94678b5b581dR108
Also, Distributions.cu have a very inefficient pattern of calling random number generation, efficient pattern is in dropout pr also.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 8]
```diff
@@ -87,11 +93,302 @@ struct TensorInfo {
...
+  public:
+    PHILOX_DEVICE_INLINE
+    explicit PhiloxWrapper(uint64 seed_lo, uint64 seed_hi, uint64 offset) {
+      philox = PhiloxRandom(seed_lo, seed_hi);
+      philox.Skip(offset / 4);
+      i = 0;
+    }
+    PHILOX_DEVICE_INLINE float operator()() {
+      if(i == 0) buf = philox();
+      uint32 ret = buf[i];
```
Comment (mcarilli) [inline torch/csrc/jit/fusion_compiler.cpp]: This is a dynamic index into a per-thread array, which will cause that array to be stored in local memory, and kill your performance by a factor of 2 or more.  Replace with the following:
```
    uint32 ret;
    switch(i)
    {
      default:
        ret = buf[0];
        break;
    case 1:
        ret = buf[1];
        break;
    case 2:
        ret = buf[2];
        break;
    case 3:
        ret = buf[3];
        break;
    }
```

File: torch/csrc/jit/fusion_compiler.cpp [hunk 9]
```diff
@@ -87,11 +93,302 @@ struct TensorInfo {
...
+      i = (i + 1) % 4;
+      const uint32 FLOAT_MASK = (1 << 24) - 1;
```
Comment (ngimel) [inline torch/csrc/jit/fusion_compiler.cpp]: Why are you throwing away 8 bits of randomness here?

File: torch/csrc/jit/fusion_compiler.cpp (hunk 6 above)
Comment (mcarilli) [reply torch/csrc/jit/fusion_compiler.cpp]: I believe that 10 or 20 estimate was specific to the case of generating Poisson-distributed random numbers.  

If you're just generating uniformly distributed floats or ints one at a time, you're manually invoking the generator, so you know exactly how many times you're calling the generator per thread, as @zdevito says.  

The original tensorflow Philox anticipates that `offset` is a multiple of 4, where `offset` is the number of 32-bit values generated so far (aka, in prior kernel launches) for this particular subsequence.  We should stick to that convention to be safe.  You're already dividing `offset` by 4 [when you construct PhiloxRandom](https://github.com/pytorch/pytorch/pull/9795/files#diff-c76c51fc543bd5257d3ae9bb400bdbd8R357) within PhiloxWrapper, which is fine.  Since `offset` here is

File: torch/csrc/jit/fusion_compiler.cpp (hunk 9 above)
Comment (zrt95) [reply torch/csrc/jit/fusion_compiler.cpp]: This is just for casting the int to floating point. We are gonna lose randomness anyway if we want to get a floating point number within [0,1)

Comment (ngimel) [reply torch/csrc/jit/fusion_compiler.cpp]: But not whole 8 bits (one quarter! of it). You might as well have a float divisor of 1.0f/2^32, that would give you higher density of (float) randoms near 0.

File: torch/csrc/jit/fusion_compiler.cpp [hunk 10]
```diff
@@ -88,11 +94,116 @@ struct TensorInfo {
 };
 )");
 
+// The reason why we used TensorFlow's philox implementation is that currently
```
Comment (ngimel) [inline torch/csrc/jit/fusion_compiler.cpp]: This comment is not accurate any more?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 11]
```diff
@@ -88,11 +94,116 @@ struct TensorInfo {
...
+    }
+
+    static const unsigned long kPhilox10A = 0x9E3779B9;
+    static const unsigned long kPhilox10B = 0xBB67AE85;
+    static const unsigned long kPhiloxSA = 0xD2511F53;
+    static const unsigned long kPhiloxSB = 0xCD9E8D57;
+  };
+
+  // Constants are picked from https://www.doornik.com/research/randomdouble.pdf
+  #define M_RAN_INVM32 2.32830643653869628906e-010
```
Comment (ngimel) [inline torch/csrc/jit/fusion_compiler.cpp]: Can we make this a floating point literal (not double) not to trigger slow double precision arithmetic?

File: torch/csrc/jit/fusion_compiler.cpp [hunk 12]
```diff
@@ -88,11 +94,116 @@ struct TensorInfo {
...
+    uint4 output;
+    uint2 key;
+    unsigned int STATE;
+    __device__ inline void incr_n(unsigned long long n) {
+      unsigned int nlo = (unsigned int)(n);
+      unsigned int nhi = (unsigned int)(n >> 32);
+      counter.x += nlo;
+      if (counter.x < nlo)
+        nhi++;
+      counter.y += nhi;
```
Comment (ngimel) [inline torch/csrc/jit/fusion_compiler.cpp]: nhi should also be checked for overflow.

**Comment by ngimel** [comment]
This looks better now, I have a more general question - do we want philox generator as a string literal, or do we want it as a header that's redistributed with binaries and thus can be supplied to nvrtc as a header, not as part of a source file? Pros of going with the header is that when THC generation is moved to use philox too (to match jit) it can reuse this header, otherwise THC will either have to use curand or have a copy-paste of what's in this PR, both of those options are not great IMO.

**Comment by zrt95** [comment]
I wasn't quite sure how we shall do the redistribute the header. I think NVRTC either need the path to the header file or the actual string literal rather than using just the include directive.

**Comment by ngimel** [comment]
I think it does need path to the header file, but pytorch should be able to provide that at runtime if it's part of pytorch binary install?

**Comment by ezyang** [comment]
I agree it's "better" for the header to be in an actual file, but to do this we have to solve some redistribution problems, where the JIT code doesn't "know" where we installed ATen/TH headers, so what file should it pass to the compiler? (You can hardcode the filepath into the binary, but congratulations, your binary is no longer relocatable). It's just generally easier to make things work if you have the string in the binary.

This isn't a fatal problem; for example, you can get the header location from Python and pass it in (like how `torch/utils/cpp_extension.py` does it). It's just more work. @zrt95, it's up to you; if you want to merge the PR as is to start with I'm OK with that.

**Comment by zrt95** [comment]
I think I'd prefer merge it as is.

**Comment by ezyang** [comment]
you have my blessing

**Comment by mcarilli** [comment]
LGTM.  I copied the Philox code and [benchmarked it](https://gist.github.com/mcarilli/1eaacb0ba201d17209ab539d20a03ace) on my Titan V (roughly equivalent to V100).  Performance is decent and I'm not observing any pesky local memory use.
